File: benchmark/autolifter/segment-tree/max-1s-p-cover.f
Config ExtraGrammar = "AutoLifter";
Config ComposeNum = 4;
Config SampleSize = 20;

Inductive List = nil Unit | cons {Int, List};
Tag = Int; Result = Compress List; NodeInfo = {Tag, Result};
Inductive SegTree = empty Unit | node {NodeInfo, SegTree, SegTree};
Inductive Operation = update {Int, Int, Tag} | query {Int, Int};
Inductive OpList = onil Unit | ocons {Operation, OpList};

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

map = \g: Int -> Int. fix (
  \f: List -> List. \xs: List.
  match xs with
    nil _ -> xs
  | cons {h, t} -> cons {g h, f t}
  end
);

solve = \merge_tag: Tag -> Tag -> Tag. \default_tag: Tag.
  \apply: Tag -> Int -> Int. \target: List -> {Int, Int}.
let get_info = \n: SegTree. match n with node {info, _, _} -> info end in
let merge_res = \x: Result. \y: Result. concat x y in
let merge = \l: SegTree. \r: SegTree.
  let linfo = get_info l in let rinfo = get_info r in
    node {{default_tag, merge_res linfo.2 rinfo.2}, l, r} in
let apply = \n: SegTree. \tag: Tag.
  match n with
    node {info, l, r} ->
      node {{merge_tag info.1 tag, map (apply tag) info.2}, l, r}
  end in
let pushdown = \n: SegTree.
  match n with
    node {info, l, r} ->
      node {{default_tag, info.2}, apply l info.1, apply r info.1}
  end in
let buildtree = fix (
  \f: Int -> Int -> List -> {List, SegTree}. \l: Int. \r: Int. \xs: List.
  if == l r then
    match xs with
      cons {h, t} ->
        {t, node {{default_tag, cons {h, nil unit}}, empty unit, empty unit}}
    end
  else let mid = / (+ l r) 2 in
    let lres = f l mid xs in
      let rres = f (+ mid 1) r lres.1 in
        {rres.1, merge lres.2 rres.2}
) in
let update_tree = \ql: Int. \qr: Int. \tag: Tag. fix (
  \f: Int -> Int -> SegTree -> SegTree.
  \l: Int. \r: Int. \n: SegTree.
  if or (> l ql) (< r qr) then n
  else if and (<= ql l) (<= r qr) then apply n tag
  else let mid = / (+ l r) 2 in
    match (pushdown n) with
      node {info, lnode, rnode} ->
        let lres = f l mid lnode in
        let rres = f (+ mid 1) r rnode in
          merge lres rres
    end
) in
let query_tree = \ql: Int. \qr: Int. fix (
  \f: Int -> Int -> SegTree -> {SegTree, Result}.
  \l: Int. \r: Int. \n: SegTree.
  if or (> l ql) (< r qr) then {n, nil unit}
  else if and (<= ql l) (<= r qr) then {n, (get_info n).2}
  else let mid = / (+ l r) 2 in
    let new_node = pushdown n in
    match new_node with
      node {info, lnode, rnode} ->
        let lres = f l mid lnode in
        let rres = f (+ mid 1) r rnode in
          {node {info, lres.1, rres.1}, merge_res lres.2 rres.2}
    end
) in
\init: List. \ops: OpList.
let len = length init in
if == len 0 then nil unit else
let is_range_valid = \l: Int. \r: Int. and (<= 1 l) (and (<= l r) (<= r len)) in
let root = (buildtree 1 len init).2 in
fix (
  \f: SegTree -> OpList -> List. \root: SegTree. \ops: OpList.
  match ops with
    onil _ -> nil unit
  | ocons {h, t} ->
    match h with
      query {l, r} ->
        if is_range_valid l r then
          let res = query_tree l r 1 len root in
            cons {(target res.2).1, f res.1 t}
        else f root t
    | update {l, r, tag} ->
        if is_range_valid l r then
          let res = update_tree l r tag 1 len root in
            f res t
        else f root t
    end
  end
) root ops;

/*Tag-related operators*/

default_tag = 100;
apply_tag = \t: Tag. \w: Int.
  if == default_tag t then w else t;
merge_tag = \x: Tag. \y: Tag.
  if == default_tag y then x else y;

/*Customized Generator*/
fit_int = \a: Int. if == (* 2 (/ a 2)) a then 0 else 1;
fit_list = fix (
  \f: List -> List. \xs: List.
  match xs with
    nil _ -> xs
  | cons {h, t} -> cons {fit_int h, f t}
  end
);
fit_ops = fix (
  \f: OpList -> OpList. \xs: OpList.
  match xs with
    onil _ -> xs
  | ocons {update {l, r, tag}, t} ->
    ocons {update {l, r, fit_int tag}, f t}
  | ocons {h, t} -> ocons {h, f t}
  end
);

/*Query-related operators*/

max1s_with_pos = fix (\f: Int -> Int -> List -> {Int, Int}. \pre: Int. \i: Int. \xs: List.
  match xs with
    nil _ ->
      let len = - i pre in
        {len, pre}
  | cons {h, t} ->
      if == h 1 then f pre (+ i 1) t
      else let len = - i pre in
        let res = f (+ i 1) (+ i 1) t in
          if >= len res.1 then {len, pre} else res
  end
) 0 0;

main = \raw_init: List. \raw_ops: OpList.
  let init = fit_list raw_init in
  let ops = fit_ops raw_ops in
    solve merge_tag default_tag apply_tag max1s_with_pos init ops;

File: ml-benchmark/autolifter/segment-tree/max-1s-p-cover.ml
type lst = Nil | Cons of int * lst
type tag = int
type result = lst
type nodeinfo = tag * result
type segtree = Empty | Node of nodeinfo * segtree * segtree
type operation = Update of int * int * tag | Query of int * int
type oplist = ONil | OCons of operation * oplist

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil -> y

let rec length x =
  match x with
  | Cons (_, t) -> length t + 1
  | Nil -> 0

let map g =
  let rec f xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (g h, f t)
  in
  f

let solve merge_tag default_tag apply_elem target =
  let get_info n =
    match n with
    | Node (info, _, _) -> info
    | _ -> failwith ""
  in
  let merge_res x y = concat x y in
  let merge l r =
    let linfo = get_info l in
    let rinfo = get_info r in
    match linfo with
    | (ltag, lres) ->
      match rinfo with
      | (_, rres) ->
        Node ((default_tag, merge_res lres rres), l, r)
  in
  let apply_node n tagv =
    match n with
    | Node (info, l, r) ->
      (match info with
       | (itag, ires) ->
         Node ((merge_tag itag tagv, map (apply_elem tagv) ires), l, r))
    | _ -> n
  in
  let pushdown n =
    match n with
    | Node (info, l, r) ->
      (match info with
       | (itag, ires) ->
         Node ((default_tag, ires), apply_node l itag, apply_node r itag))
    | _ -> n
  in
  let rec buildtree l r xs =
    if l = r then
      match xs with
      | Cons (h, t) ->
        let leaf = Node ((default_tag, Cons (h, Nil)), Empty, Empty) in
        (t, leaf)
      | _ -> (Nil, Empty)
    else
      let mid = (l + r) / 2 in
      let lres = buildtree l mid xs in
      match lres with
      | (rest1, ltree) ->
        let rres = buildtree (mid + 1) r rest1 in
        match rres with
        | (rest2, rtree) ->
          (rest2, merge ltree rtree)
  in
  let update_tree ql qr tagv =
    let rec f l r n =
      if l > qr || r < ql then n
      else if ql <= l && r <= qr then apply_node n tagv
      else
        let mid = (l + r) / 2 in
        let n2 = pushdown n in
        match n2 with
        | Node (_, lnode, rnode) ->
          let lres = f l mid lnode in
          let rres = f (mid + 1) r rnode in
          merge lres rres
        | _ -> n2
    in
    f
  in
  let query_tree ql qr =
    let rec f l r n =
      if l > qr || r < ql then (n, Nil)
      else if ql <= l && r <= qr then
        let info = get_info n in
        match info with
        | (_, ires) -> (n, ires)
      else
        let mid = (l + r) / 2 in
        let new_node = pushdown n in
        match new_node with
        | Node (info, lnode, rnode) ->
          let lres = f l mid lnode in
          (match lres with
           | (lnode2, lacc) ->
             let rres = f (mid + 1) r rnode in
             (match rres with
              | (rnode2, racc) ->
                (Node (info, lnode2, rnode2), merge_res lacc racc)))
        | _ -> (new_node, Nil)
    in
    f
  in
  fun init ops ->
    let len = length init in
    if len = 0 then Nil else
    let is_range_valid l r = 1 <= l && l <= r && r <= len in
    let root2 =
      let br = buildtree 1 len init in
      match br with
      | (_, tree) -> tree
    in
    let rec loop root ops =
      match ops with
      | ONil -> Nil
      | OCons (h, t) ->
        (match h with
         | Query (l, r) ->
           if is_range_valid l r then
             let res = query_tree l r 1 len root in
             (match res with
              | (root3, acc) ->
                let tg = target acc in
                (match tg with
                 | (a, _) ->
                   Cons (a, loop root3 t)))
           else loop root t
         | Update (l, r, tagv) ->
           if is_range_valid l r then
             let root3 = update_tree l r tagv 1 len root in
             loop root3 t
           else loop root t)
    in
    loop root2 ops

let default_tag = 100

let apply_tag t w =
  if default_tag = t then w else t

let merge_tag x y =
  if default_tag = y then x else y

let fit_int a =
  if 2 * (a / 2) = a then 0 else 1

let rec fit_list xs =
  match xs with
  | Nil -> xs
  | Cons (h, t) -> Cons (fit_int h, fit_list t)

let rec fit_ops xs =
  match xs with
  | ONil -> xs
  | OCons (op, t) ->
    (match op with
     | Update (l, r, tagv) ->
       OCons (Update (l, r, fit_int tagv), fit_ops t)
     | Query _ -> OCons (op, fit_ops t))

let rec max1s_with_pos_f pre i xs =
  match xs with
  | Nil ->
    let len = i - pre in
    (len, pre)
  | Cons (h, t) ->
    if h = 1 then max1s_with_pos_f pre (i + 1) t
    else
      let len = i - pre in
      let res = max1s_with_pos_f (i + 1) (i + 1) t in
      match res with
      | (len2, _) ->
        if len >= len2 then (len, pre) else res

let max1s_with_pos = max1s_with_pos_f 0 0

let program raw_init raw_ops =
  let init = fit_list raw_init in
  let ops = fit_ops raw_ops in
  let s = solve merge_tag default_tag apply_tag max1s_with_pos in
  s init ops